server/src/main/java/org/opensearch/env/NodeEnvironment.java

```diff
@@ -199,6 +202,8 @@ public String toString() {
 
     private final NodeMetadata nodeMetadata;
 
+    private final SetOnce<FileCache> fileCache = new SetOnce<>();
```

Comment:
I think it makes sense to have node environment delete the files/paths for the cache like it does for the other shard paths. However, I don't love the idea of wiring in a FileCache instance to NodeEnvironment because it seems to be beyond the responsibility of NodeEnvironment. What do you think of the following:

- Keep FileCacheCleaner, but change the `beforeIndexShardDeleted` event to `afterIndexShardClosed` (which I think will be triggered in all cases, but please do confirm). In that event we will remove entries from the FileCache.
- Move the path deletion logic from FileCacheCleaner into NodeEnvironment (like you have in the PR now). This means FileCacheCleaner keeps the responsibility of removing entries from the cache and NodeEnvironment gets the logic to actually delete the files.

Let me know what you think.

Reply:
> * Keep FileCacheCleaner, but change the `beforeIndexShardDeleted` event to `afterIndexShardClosed` (which I think will be triggered in all cases, but please do confirm). In that event we will remove entries from the FileCache.

IMO, if an index has been loaded, I believe that `afterIndexShardClosed` will be triggered. However, if an index hasn't been loaded, I don't think this event will be triggered, and the file cache will always be restored upon node startup. 

If we don't want to wire in a `FileCache` instance to `NodeEnvironment`, we could consider adding a new listener to `NodeEnvironment`. For instance, we could add listeners for events such as `before/afterShardDirectoryDeleted`, `before/afterIndexDirectoryDeleted`, or `onShardDirectoryDeleted`, `onIndexDirectoryDeleted`.